### PR TITLE
feat: improve route

### PR DESCRIPTION
**Pull request by LiteSun**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bug fix
- [ ] New feature provided
- [x] Improve performance

- Related issues

___
### Bugfix
- Description

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.

### CODE REVIEW

File: src/pages/Route/components/CreateStep4/CreateStep4.tsx [hunk 1]
```diff
@@ -23,19 +23,22 @@ import PluginOrchestration from '@api7-dashboard/pluginchart';
...
 
-interface Props extends RouteModule.Data {
+type Props = {
   form1: FormInstance;
   form2: FormInstance;
   redirect?: boolean;
-}
+  step3Data: RouteModule.Step3Data;
+  advancedMatchingRules: RouteModule.MatchingRule[];
+  upstreamHeaderList: RouteModule.UpstreamHeader[];
```
Comment (juzhiyuan) [inline src/pages/Route/components/CreateStep4/CreateStep4.tsx:32]: we'd better have a unified style on List variables, e.g advancedMatchingRuleList or upstramHeaders, I would prefer the former one.

File: src/pages/Route/components/CreateStep4/CreateStep4.tsx [hunk 2]
```diff
@@ -44,17 +47,18 @@ const CreateStep4: React.FC<Props> = ({ form1, form2, redirect, ...rest }) => {
       {!redirect && (
         <>
           <h2 style={style}>{formatMessage({ id: 'route.create.define.api.backend.server' })}</h2>
-          <Step2 {...rest} form={form2} disabled />
+          <Step2
+            upstreamHeaderList={rest.upstreamHeaderList}
+            form={form2}
+            disabled
+            onChange={() => {}}
```
Comment (juzhiyuan) [inline src/pages/Route/components/CreateStep4/CreateStep4.tsx:54]: ,

File: src/pages/Route/components/CreateStep4/CreateStep4.tsx [hunk 3]
```diff
@@ -44,17 +47,18 @@ const CreateStep4: React.FC<Props> = ({ form1, form2, redirect, ...rest }) => {
...
-            <PluginPage initialData={rest.data.step3Data.plugins} readonly />
+            <PluginPage initialData={rest.step3Data.plugins} readonly />
           )}
           {Boolean(Object.keys(script).length !== 0) && (
-            <PluginOrchestration
-              data={rest.data.step3Data.script.chart}
-              readonly
-              onChange={() => {}}
-            />
+            <PluginOrchestration data={rest.step3Data.script.chart} readonly onChange={() => {}} />
```
Comment (juzhiyuan) [inline src/pages/Route/components/CreateStep4/CreateStep4.tsx:61]: ,

File: src/pages/Route/components/Step1/MatchingRulesView.tsx [hunk 4]
```diff
@@ -19,11 +19,11 @@ import { Button, Table, Modal, Form, Select, Input, Space } from 'antd';
 import { useIntl } from 'umi';
 import { PanelSection } from '@api7-dashboard/ui';
 
-interface Props extends RouteModule.Data {}
-
-const MatchingRulesView: React.FC<Props> = ({ data, disabled, onChange }) => {
-  const { advancedMatchingRules } = data.step1Data;
-
+const MatchingRulesView: React.FC<RouteModule.Step1PassProps> = ({
```
Comment (juzhiyuan) [inline src/pages/Route/components/Step1/MatchingRulesView.tsx:22]: Pass?

File: src/pages/Route/components/Step1/MetaView.tsx [hunk 5]
```diff
@@ -63,9 +60,20 @@ const MetaView: React.FC<Props> = ({ data, disabled, onChange, isEdit }) => {
...
-            if (step1Data.route_group_id) {
-              onChange({ route_group_id: value });
+            if (!value) {
+              form.setFieldsValue({
+                ...form.getFieldsValue(),
+                route_group_name: '',
+              });
+              return;
             }
+            fetchRouteGroupItem(value.toString()).then((data) => {
```
Comment (juzhiyuan) [inline src/pages/Route/components/Step1/MetaView.tsx:70]: What's this value's typing?

File: src/pages/Route/components/Step2/RequestRewriteView.tsx [hunk 6]
```diff
@@ -29,15 +29,23 @@ import {
...
-interface Props extends RouteModule.Data {
-  form: FormInstance;
-}
-const RequestRewriteView: React.FC<Props> = ({ data, form, disabled, onChange }) => {
-  const { step2Data } = data;
+const RequestRewriteView: React.FC<RouteModule.Step2PassProps> = ({ form, disabled }) => {
   const [upstearms, setUpstreams] = useState<{ id: string; name: string }[]>();
-  const upstreamDisabled = disabled || !!step2Data.upstream_id;
+  const [upstreamId, setUpstreamId] = useState(form.getFieldValue('upstream_id'));
+  // TODO: need to check
```
Comment (juzhiyuan) [inline src/pages/Route/components/Step2/RequestRewriteView.tsx:37]: any solution here?

File: src/pages/Route/transform.ts [hunk 7]
```diff
@@ -18,49 +18,53 @@
...
+}: RouteModule.RequestData) => {
   const nodes = {};
-  step2Data.upstreamHostList.forEach((node) => {
+  (form2Data.upstreamHostList || []).forEach((node) => {
     nodes[`${node.host}:${node.port}`] = node.weight;
   });
 
   const upstream_header = {};
-  (step2Data.upstreamHeaderList || []).forEach((header) => {
+  (upstreamHeaderList || []).forEach((header) => {
```
Comment (juzhiyuan) [inline src/pages/Route/transform.ts:33]: could we set a default value for this list?

File: src/pages/Route/transform.ts [hunk 8]
```diff
@@ -76,34 +80,33 @@ export const transformStepData = ({
...
-  if (step3Data.plugins.prometheus) {
-    // eslint-disable-next-line no-param-reassign
-    step3Data.plugins.prometheus = {};
-  }
-
   // 未启用 redirect
   if (!redirect.uri) {
+    if (step3Data.plugins.prometheus) {
+      // eslint-disable-next-line no-param-reassign
+      step3Data.plugins.prometheus = {};
```
Comment (juzhiyuan) [inline src/pages/Route/transform.ts:108]: This logic is no longer needed because of the new PluginModule

**Review by juzhiyuan — APPROVED** [review]
Looks good to me from those codes style, but could not make sure if it's working actually, we need some test cases to cover all PRs.

File: src/pages/Route/transform.ts (hunk 8 above)
Comment (LiteSun) [reply src/pages/Route/transform.ts:108]: ok

File: src/pages/Route/Create.tsx [hunk 9]
```diff
@@ -68,80 +60,63 @@ const Page: React.FC<Props> = (props) => {
...
     form1.setFieldsValue(DEFAULT_STEP_1_DATA);
     form2.setFieldsValue(DEFAULT_STEP_2_DATA);
     setStep(1);
   };
 
+  useEffect(() => {
+    if (props.route.path.indexOf('edit') !== -1) {
+      setupRoute(props.match.params.rid);
+    } else {
+      onReset();
```
[the remark this answers is not in the document]
Comment (LiteSun) [reply src/pages/Route/Create.tsx:99]: > Why we should reset all fields when creating a route? the initial data should be empty, right?
`onReset()` is here to initialize the page data.

File: src/pages/Route/transform.ts (hunk 7 above)
Comment (LiteSun) [reply src/pages/Route/transform.ts:33]: Yep

File: src/pages/Route/Create.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (juzhiyuan) [reply src/pages/Route/Create.tsx:99]: setupXX would be better than reset

**Review by juzhiyuan — APPROVED** [review]
good catch

**Review by moonming — APPROVED** [review]

**Pull request merged**